design/rest-api-documentation.md

```diff
@@ -0,0 +1,133 @@
+
+```
+<dependency>
+    <groupId>org.eclipse.microprofile.openapi</groupId>
+    <artifactId>microprofile-openapi-api</artifactId>
+</dependency>
+```
+
+By spinning up the Keycloak server, developers should be able to send an HTTP request to http://localhost:8080/auth/openapi and get the OpenAPI document. Here is an example:
+
+```
+$ curl http://localhost:8080/auth/openapi
+---
+openapi: 3.0.1
+info:
+  title: Account REST API
+  contact:
+    name: Keycloak team
+    url: https://keycloak.org
+    email: keycloak-dev@googlegroups.com
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: v1alpha1
+servers:
+- url: /auth
+tags:
+- name: Account REST API
+  description: Account REST API operations.
+- name: OpenAPI Example
+  description: Get a text in various formats
+- name: Account REST service
+  description: REST API for the Account REST service
+paths:
+  /account/applications:
+    get:
+      tags:
+      - Account REST API
+      description: List of applications
+      parameters:
+      - name: name
+        in: query
+        schema:
+          type: string
+      responses:
+        "401":
+          description: Unauthorized, Failure
+        "500":
+          description: Error in getting the list of applications
+        "200":
+          description: Successful, List of applications
+```
+
+As we can see, with minimal effort it is possible to produce a JSON document that describes the functionalities of our API and can be used as a foundation for UI tools like Swagger to visualize and interact with our APIs. This document is served as a static file.
+
+## Swagger UI
+
+Swagger UI allows developers to visualize and generate documentation based on the OpenAPI specification. There are several ways of doing this:
+
+1. Inserting the URL in the [Swagger generator](https://generator.swagger.io/) which is available online. For example: http://localhost:8080/auth/openapi
+2. Spinning up a Docker image: 
+```
+docker pull swaggerapi/swagger-ui
+docker run -p 80:8080 swaggerapi/swagger-ui
+```   
+3. [Swagger Codegen](https://github.com/swagger-api/swagger-codegen#building). The best way to generate static HTML files for documentation.
+
+## Implementation plan
+
+1. Upgrade Keycloak distribution to WildFly 20.0.1.Final
+
+The current Keycloak distribution still use legacy feature packs from WildFly. In order to provide a subsystem xml file so that it can be used, we need the latest changes from [introduced in WildFly 20.0.1.Final](https://issues.redhat.com/browse/WFLY-13766).
+
+2. Introduce Microprofile OpenAPI subsystem in the Keycloak distributions
+3. Update the new Account REST API with the proper annotations
+4. Make sure that nothing is broken with the Quarkus distribution
+
+### Future work items
+
+The items below were included as nice to have because it's fairly easy to get these documents out of date and introduce it requires more discussion.
```

Comment:
I don't understand this comment. Do you mean if they are not generated, but rather some how hand crafted?

Reply:
@stianst I can do some updates in this paragraph. But in summary, I'd like to split the work into 2 steps if we agree:

1. Dynamic documentation. 

That's the most important thing at the moment from my point of view. And that means that changes into the code should immediately be reflected in the OpenAPI document. 

2. Static documentation

Important, but we will only generate it at each release. That means that changes between releases won't be reflected immediately and in theory, could get out of date. That's what I meant, but of course, it's not the end of the world.

The reason why I see 1 as something more important, it's because it requires changes on Keycloak that depends on WildFly and requires more testing. And also I didn't want to introduce too many changes in one go.